src/keyboard/textinput.js

```diff
@@ -199,19 +224,49 @@ var TextInput = function(parentNode, host) {
             var selection = host.selection;
             var range = selection.getRange();
             var row = selection.cursor.row;
-            selectionStart = range.start.column;
-            selectionEnd = range.end.column;
-            line = host.session.getLine(row);
 
-            if (range.start.row != row) {
-                var prevLine = host.session.getLine(row - 1);
-                selectionStart = range.start.row < row - 1 ? 0 : selectionStart;
+            // We keep 2*numberOfExtraLines + 1 lines in the textarea, if the new active row
+            // is within the current block of lines in the textarea we do nothing. If the new row
+            // is one row above or below the current block, move up or down to the next block of lines.
+            // If the new row is further than 1 row away from the current block grab a new block centered 
+            // around the new row.
+            if (row === rowEnd + 1) {
+                rowStart = rowEnd + 1;
+                rowEnd = rowStart + 2*numberOfExtraLines;
+            } else if (row === rowStart - 1) {
+                rowEnd = rowStart - 1;
+                rowStart = rowEnd - 2*numberOfExtraLines;
+            } else if (row < rowStart - 1 || row > rowEnd + 1) {
+                rowStart = row > numberOfExtraLines ? row - numberOfExtraLines : 0;
+                rowEnd = row > numberOfExtraLines ? row + numberOfExtraLines : 2*numberOfExtraLines;
+            }
+            
+            var lines = [];
+
+            for (var i = rowStart; i < row; i++) {
+                lines.push(host.session.getLine(i));
+            }
+
+            lines.push(host.session.getLine(row));
+
+            for (var i = row + 1; i <= rowEnd; i++) {
+                lines.push(host.session.getLine(i));
+            }
```

Comment:
can it be simplified to this? seems like yes
```
for (var i = rowStart; i <= rowEnd; i++) {
    lines.push(host.session.getLine(i));
}
```

Reply:
yeah it can, in a previous revision I used each component of these 3 separately but now that I don't anymore we can combine them yeah